content/guides/additional/client.md

```diff
@@ -0,0 +1,2154 @@
+
+
+##  Communications Protocol
+
+### Changes to Back End
+
+#### Urbit as Database
+
+Our app's state should retain a list of _all_ high scores it receives.  To that end, we will transition from a single high-score state to using a `(map fren score)`.
+
+```hoon
+/-  *flap, pals
+::  * * *
+::
+|%
++$  versioned-state
+  $%  state-zero
+      state-one
+  ==
++$  state-one  $:
+      %one
+      =score
+      =scores
+    ==
++$  state-zero  $:
+      %zero
+      =score
+      hiscore=score
+    ==
++$  card  card:agent:gall
+--
+```
+
+We will include an upgrade path for the old agent:
+
+```hoon
+++  on-load
+  |=  old-state=vase
+  ^-  (quip card _this)
+  =/  old  !<(versioned-state old-state)
+  ?-  -.old
+    %one   `this(state old)
+    %zero  `this(scores (~(put by scores) our.bol hiscore.old))
+  ==
+```
+
+These values will be changed when peer state changes are received using `%pals`.
+
+#### Adding Friends
+
+`%pals` is a very simple contact manager which recognizes outgoing requests, incoming requests, and mutually recognized peers.  Let's integrate knowledge of other friends (and thus the ability to maintain a leaderboard of our friends).  We will base our leaderboard on outgoing requests for simplicity.
+
+You should copy `%pals` support files over to our working desk.
+
+```sh
+# Include the %pals tooling by ~paldev.
+git clone https://github.com/Fang-/suite.git
+cp suite/lib/pals.hoon comet/flap/lib
+cp suite/sur/pals.hoon comet/flap/sur
+cp -r suite/mar/pals comet/flap/mar
+```
+
+When dealing with `%pals`, we need to maintain some list of our friends and their current known high scores (which means modifying the app's state) and dealing with sending and receiving notifications.  That is reflected in adding `scores` to the new state in the state definition in `/app/flap.hoon`.
+
+Subscriptions will all take place over the `/flap` path.  In the `++on-poke` arm, we need to issue notices along that arm with the `%flap-update` mark.
+
+```hoon
+  %flap-action
+=/  axn  !<(action vase)
+?>  =(-.axn %gain)
+?.  (gth score.axn hiscore)
+  `this(score score.axn)
+:_  this(score score.axn, hiscore score.axn)
+:~  [%give %fact ~[/flap] %flap-update !>(`update`lord+[score=score.axn fren=our.bol])]
+==
+```
+
+We also need to process incoming `%fact`s with a `%flap-update` cage in `++on-agent`:
+
+```hoon
+  %flap-update
+=/  upd  !<(update q.cage.sign)
+?>  =(-.upd %lord)
+?:  (gth (~(got by scores) fren.upd) score.upd)
+  `this
+~&  >  "%flap:  new high score {<score.upd>} from {<fren.upd>}"
+`this(scores (~(put by scores) fren.upd score.upd))
+```
+
+Changes to `%pals` come in along the path `/newpals`, so we need to watch for incoming values there.  `%pals` organizes friends to `%meet` for the first time or `%part` when they separate.  Thus we add or remove entries from our `scores.state`.
+
+```hoon
+  [%newpals ~]
+?+    -.sign  (on-agent:default wire sign)
```

Comment:
I would personally add %kick resubscribe logic for this also. You're less likely to get kicks from a local agent but it's still theoretically possible and if it happens you'll be permanently unsubscribed from pals

Reply:
Added 

```
        %kick
      :_  this
      :~  [%pass /flap %agent [src.bol dap.bol] %watch /flap]
      ==
```